### ISSUE TITLE
[issue]: Can't see nvme drive in localboot

### ISSUE THREAD

**Issue by iJebus** [ventoy/Ventoy#2917]
### Official FAQ

- [X] I have checked the official FAQ.

### Ventoy Version

1.0.99

### What about latest release

Yes. I have tried the latest release, but the bug still exist.

### Try alternative boot mode

No. I didn't try these alternative boot modes.

### BIOS Mode

Legacy BIOS Mode

### Partition Style

MBR

### Disk Capacity

32GB

### Disk Manufacturer

Kingston

### Image file checksum (if applicable)

No.

### What happened?

So my situation is something that I *think* should work, but am not sure and hopefully can get clarified here.

I have an old HP N40L Microserver on which I'm trying to install TrueNAS, to a 500GB M.2 SSD via an nvme-to-pcie adapter. The nvme isn't visible to the BIOS natively, but using Ventoy, I'm able to load the TrueNAS installer ISO. Through this method, the installer is able to see the nvme drive; if I try without using Ventoy, the disk isn't visible as an install target.

Of course, after the install completes, the nvme drive still isn't visible to the BIOS and so TrueNAS won't boot. I was hoping to use Ventoy to boot from the nvme but... this doesn't seem to be working.

The nvme drive isn't visible under the Browse menu and trying to boot the first disk via the the Localboot menu results in a flashing prompt on a black screen. I'm thinking that nvme drivers aren't loaded in ventoy at this point, would that be correct?

I think it's close to working but... :shrug: I've spent too many hours trying to resolve this so it's time to ask for advice.

**Comment by steve6375** [comment]
Use a nvme m.2 to USB 3 adaptor.

**Comment by steve6375** [comment]
Presumably, you are talking about a legacy/BIOS boot, not a UEFI boot as you said the HP N40L is old?
Running TrueNAS from a m.2 - USB adapter would be possible if TrueNAS runs from memory (enclosures are approx $30). Initial USB 2 booting may be a bit slow.
https://amzn.to/4cMrXog

To BIOS boot from Ventoy. you would need to find and add a grub2 nvme driver to Ventoy which is difficult if you are not using UEFI (there seems to be a way around it if using UEFI by using the clover nvme uefi driver and booting to Clover).
see
https://superuser.com/questions/1782009/can-grub-access-an-nvme-drive-when-the-bios-lacks-support-for-it

**Comment by iJebus** [comment]
TLDR; I got it working with Clover directly. Thanks for your replies.

---

Yep, legacy BIOS. Thanks for the linking to Clover, it was something I actually tried before Ventoy but wasn't having much luck with. Turns out, with much more forum reading, newer versions of Clover are essentially broken in their legacy BIOS support. So finding and using an old version, I was able to see the nvme drive... sometimes. 

Turns out for whatever reason (presumably, speed), the nvme drive isn't getting picked up by the time BIOS finishes from a *cold* boot. If  I do a reset, the drive is detected. I'm happy to live with this even if it's tedious. Given the machine is a NAS that should be running 24/7 and is connected to a UPS, cold boots should be very rare.

**Comment by deepvision7** [comment]
I have a similar situation with hardware 
- An old MB with UEFI, but without native NVMe support
- A Samsung M.2 disk
- An M.2 to PCIe adapter card
- VHD image on the NVMe drive
 
I was able to boot into VHD image with Clover thumb drive. But I had to disable Secure Boot in BIOS to allow it to load.
Since Ventoy supports Secure Boot by default is there any way to also load NvmExpressDxe.efi driver from Ventoy to make it see the NVMe drive ?

Thank you in advance.
